### PR TITLE
train:test:val split implemented

### PR DESCRIPTION
**Pull request by Lakshay-G**

### CODE REVIEW

**Review by Sehmimul — CHANGES_REQUESTED** [review]
Looks good on my end, but requested small change.

File: src/dataloader/dataloader.py [hunk 1]
```diff
@@ -57,8 +57,13 @@ def __init__(self, data_path, total_image_num, train=True):
             now selecting randomly generated n number of images. n is mentioned in total_images in params.json
             mulitplier 0.3 mentioned to make sure we select 0.3 * n images from all testing images
             '''
-            self.dataset = random_image_selection(
```
Comment (Sehmimul) [inline src/dataloader/dataloader.py:60]: Why is there if elif inside else? We can probably have just if elif else

Comment (Lakshay-G) [reply src/dataloader/dataloader.py:60]: If the dataset is not train then there are less transformations for val and test images. They are all contained in the else block, so it made more sense to me to have it there instead of having an if elif else block, just a shorter code version

Comment (vishnu-m77) [reply src/dataloader/dataloader.py:60]: We need the code to be understandable so having more lines of code is preferred if it makes it more readable.

Comment (Lakshay-G) [reply src/dataloader/dataloader.py:60]: I've added comments for it. If needed, I can add more lines of code. Let me know once you look at it.

File: README.md [hunk 2]
```diff
@@ -24,6 +24,22 @@ dataset/aptos/
...
+
+`test_batch_size`: Currently set to 2 as the inference can happen only for that, however higher batch_size ~ 25 is recommended.
+
+`timestep`: Currently experimented with 500, 80, 60, 50. Optimally ~ 60 is recommended.
+
+`num_classes`: Selected as 5 for 5 classes - 0, 1, 2, 3, 4. Don't change it unless you change the dataset.
+
+`include_guidance`: Ensures that DCG priors are used in diffusion. `true` indicates using DCG priors for diffusion, `false` indicates not using DCG priors for diffusion.
+
+`weight`: For a detailed description look at Diffusion section of README.
```
Comment (vishnu-m77) [inline README.md:41]: Can you add a link to the diffusion section instead of just mentioning it?

File: README.md [hunk 3]
```diff
@@ -42,11 +58,14 @@ dataset/aptos/
...
 `diffusion.py`: Contains code for forward and reverse diffusion.
 
-A class for `weighted_loss` has been implemented which takes `weight` as a parameter that is passed from the diffusion parameters `params["diffusion"]["weight"]`. There are three types of loss: 
+A class for `weighted_loss` has been implemented which takes `weight` as a parameter that is passed from the diffusion parameters `params["diffusion"]["weight"]`. There are three types of loss:
+
 - Unweighted MMD loss (`weight=None`)
 - Weighted MMD loss with the inverse of number of images in each class (`weight=n`)
 - Weighted MMD loss with the square root of the inverse of number of images in each class (`weight=sqrt`)
 
+Weights are calculated currently using total images. There are [1805, 370, 999, 193, 295] images for classes [0, 1, 2, 3, 4] respectively. Thus the weight is currently static and set to `[1/1805, 1/370, 1/999, 1/193, 1/295]`. It can be made dynamic by finding the total number of images for each batch and then setting weights based on number of images of each class in the batch. This is an extension of the project.
```
Comment (vishnu-m77) [inline README.md:67]: We don't need the comment about dynamic weights. It does not make sense to implement it.

File: main.py [hunk 4]
```diff
@@ -72,7 +75,8 @@
         # Initialize DCG
         dcg = dcg_module.DCG(dcg_params)
         # Trains DCG and saves the model
-        dcg_module.train_DCG(dcg, dcg_params, train_loader, val_loader=val_loader)
+        dcg_module.train_DCG(dcg, dcg_params, train_loader,
```
Comment (vishnu-m77) [inline main.py:78]: This feels  unnecessary. Why add a newline with no purpose?

File: src/dataloader/dataloader.py [hunk 5]
```diff
@@ -16,19 +16,19 @@
 
 
 class APTOSDataset(Dataset):
-    def __init__(self, data_path, total_image_num, train=True):
+    def __init__(self, data_path, total_image_num, type='train'):
```
Comment (vishnu-m77) [inline src/dataloader/dataloader.py:19]: What values can this `type` parameter take? 
Is it included in the `params.json` file?

File: src/dataloader/dataloader.py [hunk 6]
```diff
@@ -130,16 +136,25 @@ def __init__(self, config, database_name="APTOS"):
         self.train_batch_size = config["train_batch_size"]
         self.test_batch_size = config["test_batch_size"]
         self.valid_batch_size = config["valid_batch_size"]
-        self.total_image_num = config["num_images"]
+        '''
+        based on issue #19, minimum total images are set to 1000. if config["num_images"] > 1000,
```
Comment (vishnu-m77) [inline src/dataloader/dataloader.py:140]: I added a comment in the issue that this minimum value should be much smaller.

File: src/dataloader/dataloader.py (hunk 1 above)
Comment (Lakshay-G) [reply src/dataloader/dataloader.py:60]: If this looks good @vishnu-m77 can you resolve the comment?

File: README.md (hunk 2 above)
Comment (Lakshay-G) [reply README.md]: Added, if you think it looks good, can you resolve the comment?

File: README.md (hunk 3 above)
Comment (Lakshay-G) [reply README.md:67]: I think it makes sense, @Sehmimul said it might be a different approach to applying weights. I think it's worth mentioning here for future work. If we don't want it here, we can have a future work section in README, I can mention it there instead.

File: main.py (hunk 4 above)
Comment (Lakshay-G) [reply main.py:73]: VS code extension added it. I tried to reverse it, it won't save without adding a line. Sorry

File: src/dataloader/dataloader.py (hunk 5 above)
Comment (Lakshay-G) [reply src/dataloader/dataloader.py:19]: We don't need to mention about `type` in `params.json`. However I will add about it in the README Dataloader section. `type` can take `train`, `val` or `test` values.

File: src/dataloader/dataloader.py (hunk 6 above)
Comment (Lakshay-G) [reply src/dataloader/dataloader.py:140]: Added, if you think it looks good, can you resolve the comment?

Comment (vishnu-m77) [reply src/dataloader/dataloader.py:140]: The comments are unedited.

**Review by vishnu-m77 — CHANGES_REQUESTED** [review]

File: src/dataloader/dataloader.py (hunk 6 above)
Comment (Lakshay-G) [reply src/dataloader/dataloader.py]: My bad, added that. Can you resolve the comments and merge?

File: README.md (hunk 3 above)
Comment (Lakshay-G) [reply README.md:67]: Can you resolve this?

Comment (vishnu-m77) [reply README.md:67]: Using dynamic weights here does not make sense. I will resolve it after the last two sentences are removed.

**Review by vishnu-m77 — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (Lakshay-G) [reply README.md:67]: @Sehmimul can you review it once you have time?

Comment (Sehmimul) [reply README.md:67]: Okay let's add a potential extension section and add that dynamic loss part. The sentence  as it is incorrect as some batches can have no images of a particular label. Sentence, needs to be explained properly.

**Review by vishnu-m77 — APPROVED** [review]

**Review by Sehmimul — APPROVED** [review]
Approved! Nice work!
